[CA-1385] A11y: Accessible Menu Buttons Popups

andrewmarcus opened the pull request:
This PR is based on the #2512 branch, which hasn't yet been merged, since I needed the keyboard navigation component.

In this PR, I create a new `MenuTrigger` component, which wraps `PopupTrigger` and properly sets up a menu in a dialog box with all the proper roles and relationships.

Some accessibility features:
* The `PopupTrigger` component now allows different roles to be used in the modal: 'dialog', 'menu', 'listbox', etc. The modal itself has the indicated role, and the `aria-haspopup` attached to the trigger button has the matching value.
* The label attached to the trigger button will now also be used as a label for the dialog box
* Users can use the up, down, home and end arrow keys to navigate the menu options

I moved `MenuButton` and associated functionality into the `PopupTrigger.js` file. In addition, I updated all applicable instances of `PopupTrigger` in the codebase to use `MenuTrigger` instead and get the new functionality.

andrewmarcus committed:
Moved TabBar and SimpleTabBar into their own file and updated references
Added arrow key navigation support
Adjusted ARIA roles for both types of tabs
Adjusted integration test to identify navigation tabs properly

andrewmarcus committed:
Refactored the keyboard-nav helpers to use a single ref and determine children via the DOM rather than via React. This seems to perform just as well and is more flexible

andrewmarcus committed:
Pulling in a slightly more expansive Xpath rule for finding clickable elements

andrewmarcus committed:
Bugfixes in SimpleTabBar, fixed padding/margin so the focus regions look correct

andrewmarcus committed:
Merge remote-tracking branch 'origin/dev' into a11y/awm-tabbar

# Conflicts:
#	src/components/common.js

andrewmarcus committed:
Merged in new tab style properties from dev

andrewmarcus committed:
Cleaned up another merge conflict

andrewmarcus committed:
Merge branch 'dev' into a11y/awm-tabbar

andrewmarcus committed:
Create a new MenuTrigger component, which wraps PopupTrigger and properly setups a menu in a dialog box
* Moved MenuButton and associated functionality into the PopupTrigger.js file
* Copied in some ARIA properties from the awm-workspace-page branch, which I plan to remove from there
* Allow different roles to be used in the modal: 'dialog', 'menu', 'listbox', etc
* Use the trigger button to label the dialog/menu box
* Updated all instances of PopupTrigger with MenuButtons to use MenuTrigger instead

andrewmarcus committed:
Added more ARIA properties to the modal:
* `aria-modal: true` because we're using a FocusTrapper
* `aria-orientation: vertical` to indicate up and down arrow keys will work

s-rubenstein commented on src/components/PopupTrigger.js:
```diff
@@ -51,21 +59,32 @@ const PopupTrigger = Utils.forwardRefWithName('PopupTrigger', ({ content, childr
+      popupProps: {
+        role,
+        'aria-labelledby': labelledby,
+        ...props
```
Passing props here as opposed to a level up is technically a functional change. I would be worried about props not relating to the popup that should be passed in getting lost. However, it looks like we weren't using the general props there anyway, so it might be fine? Just worth double checking.

andrewmarcus replied:
So the problem with just pulling in ...props in `Popup` was that it also grabs all of the internal `onClickOutside` properties for some reason. I couldn't just stick them straight into the `div`; I was getting console errors complaining about things like `stopPropagation` not being a valid DOM property. So I did it this way instead to only pass in the ones we actually care about going into the DOM, like ARIA properties.

But now that I think about it, you're right! Properties `Popup` needs, like `side`, need to be passed in one level further up since they're not explicitly pulled out in `PopupTrigger`.

I think I can either expose `popupProps` as a parameter in `PopupTrigger`, or pull out the `side`, `target` and `onClick` properties explicitly.

andrewmarcus replied:
It looks like the only unnamed prop which could possibly be passed through was `side`, so I just pulled that out here. I decided not to push them into `popupProps` inside the `PopupTrigger`, because it could change existing calls, and then it's not clear what should be in versus out. For example, I'm pulling `role` out explicitly because I'm using it elsewhere, however it also gets attached to the DOM node so it should by that logic be in popupProps instead. Better to just leave it all in props and do the right thing. :)

andrewmarcus replied:
Nevermind! Users of the `PopupTrigger` can also pass in properties to be consumed by the `onClickOutside` functionality. I forgot about that. I couldn't find anywhere that's using that, but it should be possible.

I've moved the properties back into `popupProps` inside the `PopupTrigger`. At least now it's consistent and clear what those particular properties will be used for...

s-rubenstein commented on src/pages/workspaces/workspace/Notebooks.js:
```diff
@@ -21,7 +21,9 @@ import { reportError, withErrorReporting } from 'src/libs/error'
 import * as Nav from 'src/libs/nav'
 import { notify } from 'src/libs/notifications'
-import { appIsSettingUp, convertedAppStatus, currentApp, currentAttachedDataDisk, getGalaxyCost, isCurrentGalaxyDiskDetaching } from 'src/libs/runtime-utils'
+import {
```
Nit: My preference would be to either put these on the line with the curly braces, as they were before, or newline each of them/put a newline in the middle, so we are making better use of the improved horizontal space.

andrewmarcus replied:
Sorry, I ran Intellij's "Organize Imports" script and it did this. I'll revert it.

s-rubenstein reviewed (approved):
LGTM, a few changes but nothing blocking.

andrewmarcus committed:
Merge branch 'dev' into a11y/awm-tabbar

andrewmarcus committed:
Because ...props is now being passed directly to the DOM element, ensure that any additional props are passed through as before. This seems to only apply to the `side` property, so I've explicitly pulled that out

andrewmarcus committed:
Explicitly pull out the popupProps from the other properties

andrewmarcus committed:
PR whitespace change

andrewmarcus committed:
Merge branch 'dev' into a11y/awm-menubuttons

# Conflicts:
#	src/pages/billing/Project.js

andrewmarcus committed:
Merge branch 'dev' into a11y/awm-tabbar

andrewmarcus committed:
Updated another reference after merge from dev

andrewmarcus committed:
Merge branch 'a11y/awm-tabbar' into a11y/awm-menubuttons

andrewmarcus committed:
Fixed issue attempting to extract a key from a non-existent input object

andy7i reviewed (commented):
I'll try to review this PR by today. In the meantime, could you look into this? I'm seeing some errors when I try to run this branch locally.

I ran

```
yarn install
yarn start
```
and saw:
```

Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db
```
before it errored with:

```
Failed to compile
./src/components/FooterWrapper.js
Error: [BABEL] /Users/ajang/Firecloud/terra-ui/src/components/FooterWrapper.js: Cannot find module '@babel/plugin-syntax-class-properties'
Require stack:
...
```

zarsky-broad commented:
@andy7i try to `rm -rf node_modules` and then install again.

andy7i commented on src/components/keyboard-nav.js:
```diff
@@ -85,9 +85,9 @@ export const VerticalNavigation = ({ children, ...props }) => h(ArrowKeyNavigati
         focusOn(i + 1); break
-      case 'Home': case 'PageDown':
+      case 'Home': case 'PageUp':
         focusOn(0); break
-      case 'End': case 'PageUp':
+      case 'End': case 'PageDown':
```
not sure where these buttons are on my mac keyboard -- haven't tested this

zarsky-broad replied:
Fn + up/down arrow

andy7i reviewed (approved):
looks great! verified new aria-haspopup on several types of roles and everything seemed good in manual testing as well

andrewmarcus committed:
Merge branch 'dev' into a11y/awm-tabbar

# Conflicts:
#	src/pages/billing/Project.js
#	src/pages/workspaces/List.js

andrewmarcus committed:
Added an aria-label to the SimpleTabBar added to the Project.js page

andrewmarcus committed:
Merge branch 'a11y/awm-tabbar' into a11y/awm-menubuttons

# Conflicts:
#	src/components/group-common.js
#	src/pages/workspaces/List.js

andrewmarcus committed:
Cleaned up merge conflicts, and applied MenuTriggers to the new Analysis section

andrewmarcus committed:
Merge branch 'dev' into a11y/awm-menubuttons

# Conflicts:
#	src/components/common.js
#	src/components/keyboard-nav.js
#	src/pages/workspaces/List.js
#	src/pages/workspaces/workspace/WorkspaceContainer.js

dmohs commented on src/components/PopupTrigger.js:
```diff
@@ -87,3 +110,36 @@ export const InfoBox = ({ size, children, style, side, tooltip, iconOverride })
+      style: {
+        display: 'flex', alignItems: 'center',
+        fontSize: 12, minWidth: 125, height: '2.25rem',
+        color: disabled ? colors.dark(0.7) : undefined,
+        padding: '0.875rem',
+        cursor: disabled ? 'not-allowed' : 'pointer'
```
```suggestion
        padding: '0.875rem',
        ...(disabled ? { color: colors.dark(0.7), cursor: 'not-allowed' } : { cursor: 'pointer' })
```

dmohs reviewed (approved):
With @petesantos and @zarsky-broad.

zarsky-broad commented on src/components/PopupTrigger.js:
```diff
@@ -20,27 +22,35 @@ const styles = {
+export const Popup = onClickOutside(function({ id, side = 'right', target: targetId, onClick, children, popupProps = {} }) {
+  // We're passing popupProps here rather than just props, because ...props also includes lots of internal onClickOutside properties which
+  // aren't valid to be dropped on a DOM element.
+  Utils.useConsoleAssert('aria-label' in popupProps || 'aria-labelledby' in popupProps, 'In order to be accessible, Popup needs a label')
```
We don't tend to use `in`, so to match other places:
```suggestion
  Utils.useConsoleAssert(popupProps['aria-label'] ||  popupProps['aria-labelledby'], 'In order to be accessible, Popup needs a label')
```

andrewmarcus replied:
Ok, sounds good. I think I had been trying to account for someone intentionally setting the `aria-label` to `''`, but that should probably also throw a warning. I'll clean all of these useConsoleAssert statements in a new PR anyway.

andrewmarcus committed:
Update src/components/PopupTrigger.js

Clean up of conditional styling logic

Co-authored-by: David Mohs <dmohs@broadinstitute.org>

andrewmarcus committed:
Update src/components/PopupTrigger.js

PR comment: replacing `in` property existence tests with property dereferencing, for consistency

Co-authored-by: Isaac Zarsky <zarsky@broadinstitute.org>

andrewmarcus committed:
Cleaned a linting error
